src/xtm_api.cc

```diff
@@ -0,0 +1,359 @@
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg, unsigned flags)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	assert((flags & (~XTM_QUEUE_PUSH_VALID_FLAGS)) == 0);
+	if ((flags & XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_producer_should_be_notified, true,
+			 __ATOMIC_SEQ_CST);
+
+	/*
+	 * It is necessary to explain why we are trying to push and then set
+	 * is_producer_should_be_notified flag and then tries to push again!
+	 * The fact is that consumer thread could invoke functions and checked
+	 * is_producer_should_be_notified, BEFORE it was set in this function.
+	 * In this case we try to push again (consumer thread freed space in
+	 * queue in this case).
+	 */
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+error:
+	errno = ENOBUFS;
+	return -1;
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+unsigned
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.end();
+	return cnt;
+}
+
+int
+xtm_queue_push_ptr(struct xtm_queue *queue, void *ptr, unsigned flags)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.ptr = ptr;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	assert((flags & (~XTM_QUEUE_PUSH_VALID_FLAGS)) == 0);
+	if ((flags & XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_producer_should_be_notified, true,
+			 __ATOMIC_SEQ_CST);
+
+	/** See comment about this in xtm_queue_push_fun. */
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+error:
+	errno = ENOBUFS;
+	return -1;
+}
+
+unsigned
+xtm_queue_pop_ptrs(struct xtm_queue *queue, void **ptr_array,
+		   unsigned ptr_array_count)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	void **ptr_array_begin = ptr_array;
+	void **ptr_array_end = ptr_array + ptr_array_count;
+
+	iter.begin(&queue->queue);
+	while(ptr_array < ptr_array_end &&
+	      (xtm_msg = iter.read()) != nullptr) {
+		*ptr_array = xtm_msg->ptr;
+		++ptr_array;
+	}
+	iter.end();
+	return ptr_array - ptr_array_begin;
+}
+
+int
+xtm_queue_consume(int fd)
+{
+	unsigned char tmp[XTM_PIPE_SIZE];
+	ssize_t read_bytes;
+
+	while ((read_bytes = read(fd, tmp, sizeof(tmp))) < 0 && errno == EINTR)
+		;
+	if (read_bytes < 0 && errno != EAGAIN && errno != EWOULDBLOCK)
+		return -1;
+	return 0;
+}
+
+bool
+xtm_queue_get_reset_was_full(struct xtm_queue *queue)
+{
+	return __atomic_exchange_n(&queue->is_producer_should_be_notified,
```

Comment:
Please, add an explanation somewhere in the comments why do you use barriers with this flag even though it is accessed and changed from one thread only (producer).

Reply:
Fixed